### PR TITLE
Add swapChildViews to NextCollectionView

### PR DESCRIPTION
**Pull request by paulfalgout**
- Adds `DomApi.swapEl` similar to `DomApi.replaceEl`
- Adds `children.hasView` to `NextCollectionView`

When looking into `js-framework-benchmark` I found there was no real performant way to swap to views in a collectionview.

I'm not sure how I feel about this change, but it does seem rather difficult to achieve outside of such a solution.

There are a couple of things to consider here.  Not all collectionViews must have a collection... while you could still store data on the views and use the `viewComparator` somehow, that is a lot of work if all you want to do is swap them.

If the collectionView does have a collection and is sorting based on the collection _or_ other `viewComparator` even, then this API becomes decidedly less useful as the state of the view order is also managed elsewhere.. however the same could be said for `addChildView`

Before I do any testing or documentation, I thought I'd ask for :eyes:

### CODE REVIEW

File: src/next-collection-view.js [hunk 1]
```diff
@@ -577,6 +577,17 @@ const CollectionView = Backbone.View.extend({
     this.Dom.appendContents(this.el, els, {_$el: this.$el});
   },
 
+  swapChildViews(view1, view2) {
+    if (!this.children.hasView(view1) || !this.children.hasView(view1)) {
```
Comment (rafde) [inline src/next-collection-view.js]: This is nice for drag and drop view swapping or something similar.
Does this need something like a DOM parent check for both views to make sure they are being swapped from one dom parent to another? 
Or can this support something like, view1 has no parent but will swap with view2 which has a parent. Not sure what will happen with view2. I guess the person using the function will have to figure how to handle that.

Comment (paulfalgout) [inline src/next-collection-view.js]: The Dom code actually assumes that the first view has a parent..  they both have to be children of the collection, but technically I guess at the moment the 2nd could be detached?  that gets a little weird.  Guess I could check they're both `_isAttached`..

Comment (rafde) [inline src/next-collection-view.js]: I'd put more trust in `el2.parentNode` than `_isAttached`

File: src/config/dom.js [hunk 2]
```diff
@@ -53,6 +53,24 @@ export default {
...
+      return;
+    }
+
+    const parent = el1.parentNode;
+
+    if (!parent) {
+      return;
+    }
+
+    const next1 = el1.nextSibling;
```
Comment (rafde) [inline src/config/dom.js]: what if you are swapping an only sibling?

Comment (paulfalgout) [inline src/config/dom.js]: if `nextSibling` is `null` `insertBefore` will put it at the end

File: src/next-collection-view.js (hunk 1 above)
Comment (paulfalgout) [inline src/next-collection-view.js]: maybe so.. but I think we'd have to check the `parentNode` in this function so that the children don't also swap.  We'd also have to document clearly that only attached children can be swapped, but I think that probably makes sense.

Comment (rafde) [inline src/next-collection-view.js]: I think it'd be nice to be able to swap out views that don't belong to the same parent. But an issue I foresee is swapping views and the collection view has a filter that runs and it creates some type of unintended side affect.

Comment (paulfalgout) [inline src/next-collection-view.js]: yeah I think you'll only be able to swap attached views.

we can handle two parents in `swapEl`, but that'll be a bit weird for a single collectionview to have multiple parents.. but I guess if you've overridden attachHtml it is possible

Comment (paulfalgout) [inline src/next-collection-view.js]: if you're "swapping" views between 2 collectionviews, what you're actually doing is removing a view from one collection and adding it to another.. and that's what you should do in that case.

Comment (blikblum) [inline src/next-collection-view.js]: I'm ok with the feature but allowing to swap between to different collection views will complicate the things

**Comment by rafde** [comment]
This pr looks good. Just need test coverage and it'll be good

**Comment by paulfalgout** [comment]
@marionettejs/marionette-core ok so I've finished the original concept with code coverage, however I've added a 2nd commit that covers an edge case that I have not tested.  Additionally I still need to do documentation.  However I need thumbs one way or the other on the edge case.

**Comment by paulfalgout** [comment]
Yep, I think the docs may depend a little on the consensus regarding refiltering... either by documenting the behavior or the lack their of for the edge case

**Comment by paulfalgout** [comment]
@marionettejs/marionette-core could really use some eyes on the 2nd commit.  If it gets :thumb:s I'll flush out the docs and tests.

**Comment by JSteunou** [comment]
:+1: but I'm still missing the use case

**Comment by rafde** [comment]
👍

**Comment by paulfalgout** [comment]
Ok just working on docs and then this should be ready.
